tests/test_timing.py

```diff
@@ -0,0 +1,545 @@
+from dmutils import timing
+
+from collections import OrderedDict
+from contextlib import contextmanager
+from functools import lru_cache
+from itertools import chain, product
+import json
+import logging
+import mock
+from numbers import Number
+import random
+import re
+import sys
+import time
+
+from flask import request
+import pytest
+import six
+
+
+class MalleableAny:
+    def __init__(self, condition):
+        self._condition = condition
+
+    def __eq__(self, other):
+        return self._condition(other)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._condition})"
+
+    def __hash__(self):
+        return None
+
+
+class AnySupersetOf(MalleableAny):
+    def __init__(self, subset_dict):
+        self._subset_dict = subset_dict
+        super().__init__(lambda other: self._subset_dict == {k: v for k, v in other.items() if k in self._subset_dict})
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._subset_dict})"
+
+
+class AnyStringMatching(MalleableAny):
+    _cached_re_compile = staticmethod(lru_cache(maxsize=32)(re.compile))
+
+    def __init__(self, *args, **kwargs):
+        self._regex = self._cached_re_compile(*args, **kwargs)
+        super().__init__(lambda other: isinstance(other, six.string_types) and self._regex.match(other))
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._regex})"
+
+
+@contextmanager
+def mock_time_functions():
+    with mock.patch.multiple("time", perf_counter=mock.DEFAULT, process_time=mock.DEFAULT) as time_mocks:
+        # to simulate the behaviour of these time functions, we create them all as closures capturing these common
+        # "state" variables tracking the apparent pseudo-time according to each sampling function
+        perf_counter_state = random.uniform(1000, 10000)
+        process_time_state = random.uniform(1000, 10000)
+
+        # the sampling functions just return the current state
+        time_mocks["perf_counter"].side_effect = lambda: perf_counter_state
+        time_mocks["process_time"].side_effect = lambda: process_time_state
```

Comment:
What's the difference between using `side_effect` with a lambda here, and just setting `return_value`?